### DISCUSSION TITLE
Availability Metric - count the number timeotu queries

### DISCUSSION THREAD

**Discussion by flz** [dmachard/DNS-collector#189] [General]
I've been trying to wrap my head around how to create an availability metric based on what's already available.

The first thing that came to mind was to do something like that:

```
sum(rate(dnscollector_packets_total{op_name="CLIENT_RESPONSE"}[$__rate_interval])) / sum(rate(dnscollector_packets_total{op_name="CLIENT_QUERY"}[$__rate_interval]))
```

The issue with the above is that, because of the delay between query and response, you end up with a percentage higher than 100% (see below).

<img width="810" alt="Screen Shot 2022-12-04 at 8 55 14 PM" src="https://user-images.githubusercontent.com/48219/205552560-74da80d7-d396-4794-be2a-56b717fdd53c.png">

What I've been thinking of doing instead is changing the way the cache is working by adding a counter for entries evicted by the `query-timeout` and change the Get function to be more of a "pop" (get+delete). What the counter would indicate is the number queries for which no response was seen within `query-timeout seconds. 

My problem is then passing the counter to the loggers. I could add a new field/struct in DnsMessage and periodically pull the number from the cache but it seems a bit of a hack since that counter wouldn't related to the DnsMessage it would be a part of.

Thoughts?

**Comment by dmachard** [comment]
The current cache is very basic, I like the idea to counting the number of queries evicted by the query-timeout parameter,
but The DnsMessage must not contain the counter. 

I have some proposals:
- Create a global  shared cache available from loggers, so prometheus logger will be able to read what happened in the cache and create specific metrics for it. 
- another more complex option will be to create a new section in config file to instantiate one or more caches associated to collectors or loggers, in the same way the loggers will be able to do specific things on the cache.

**Comment by dmachard** [comment]
Can you add a Feature Request for that ? Metric to count the number queries for which no response was seen within `query-timeout seconds.

**Comment by dmachard** [comment]
resolved in release v0.30.0
